demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/TokenRevokedFilter.java

```diff
@@ -29,25 +31,32 @@ public String name() {
     @Override
     public void filter(HttpRequest request, HttpResponse response, PluginFilterChain chain) {
         logger.debug("Received a new request");
-        sidFilter(request, response);
-        chain.filter(request, response);
+        sidFilter(request, response)
```

Comment:
It's possible that `chain.filter(request, response)` will be called AFTER (bad case) or BEFORE (perfect case) the "PluginFilterChain" had completed

+1 :beer: if it works

Reply:
It's callback based, it is fine!
That's Netty under the hood it feels like home.

Reply:
> It's callback based, it is fine! That's Netty under the hood it feels like home.

Could you refer to any document that prove this "callback"? I agree with Tung. Code we put in reactor could be executed after response is sent to client.

Reply:
You are tiring me, so deeply, some days.

Reply:
Because implementing a callback interface:

 - > it's clear that though asynchronous the callbacks are still executed in order
 - > And sending the response to the client would trivially be implemented in the very last filter.

Reply:
I have tried running apisix with this new reactive code and saw that when redis down, client immediately received response after sending jmap request although I set redis timeout more than 10s ----> I don't think there is any "callback" here.

Reply:
Isn't it even better?

Driver might detect "it's down" and not send the request in the first place?

Can we validate logout works?

Reply:
I have tried again with redis up. I revoked the token but client still received code 200 after sending jmap request (it is supposed to be 401).

Reply:
Chain works

And is compatible with async code

![image](https://github.com/user-attachments/assets/02df15e6-feee-4021-b2fb-619b3c9822a2)

Reply:
When the "subscribe" method in TokenRevokedFilter class is executed, APISIX understands that all filters has completed and returns the response to client before running reactive code.

Reply:
I have written a test to prove that the client could receive response before reactive code is executed.

https://github.com/hungphan227/tmail-backend/commit/5fe214c11d736c9c49b46f0a2b687dc66002ad5b

The above test pass with the old code but fail with the reactive code.

Reply:
THen let's patch APisix!

Reply:
https://github.com/apache/apisix-java-plugin-runner/pull/313

Please qualify the new plugin with the new plugin runner ;-)

Reply:
Yeah with the updated plugin runner, the reactive code works well.

I have tested case token revoked --> pass
I have conducted performance test like this https://github.com/linagora/tmail-backend/issues/1247#issuecomment-2545209866 --> The result is good. The latency is always around 5s (I set timeout to 5s).

Regarding updating the plugin runner, I wonder if there is anyone reviewing and approving the PR. In case the repository is no longer maintained, I suggest we fork, create our own repository for the plugin runner (as well as CI/CD)

Reply:
> Regarding updating the plugin runner, I wonder if there is anyone reviewing and approving the PR.

It is Christmass time.

Be patient.

We can package, build and deploy the PR for the time being.

Can we do a JMAP perf test on top of this change?
Can we publish the updated Apisix docker image with both changes to the runner and to the plugin?

Reply:
Apisix docker image: hungphan227/apisix:3.9.1-debian-javaplugin

Reply:
It is complicated to update staging env to run jmap perf test with apisix. Therefore, I will update james-gatling and run it locally with tmail-backend memory version.